README.md
# HipHop for PHP

HipHop is a high performance PHP toolchain. Currently supported platforms are Linux and FreeBSD. There is no OS X support.

* [Developer Mailing List](http://groups.google.com/group/hiphop-php-dev)
* [Wiki](http://wiki.github.com/facebook/hiphop-php)
* [Issue Tracker](http://github.com/facebook/hiphop-php/issues)

## Required Packages

The latest information is available on the [wiki](http://wiki.github.com/facebook/hiphop-php/building-and-installing)

* cmake *2.6 is the minimum version*
* g++/gcc *4.3 is the minimum version*
* Boost *1.37 is the minimum version*
* flex
* bison
* re2c
* libmysql
* libxml2
* libmcrypt
* libicu *4.2 is the minimum version*
* openssl
* binutils
* libcap
* gd
* zlib
* tbb *Intel's Thread Building Blocks*
* [Oniguruma](http://www.geocities.jp/kosako3/oniguruma/)
* libpcre
* libexpat
* libmemcached

The following packages have had slight modifications added to them. Patches are provided and should be made against the current source copies.

* [libcurl](http://curl.haxx.se/download.html)
* src/third_party/libcurl.fb-changes.diff
* [libevent 1.4](http://www.monkey.org/~provos/libevent/)
* src/third_party/libevent-1.4.13.fb-changes.diff	OR src/third_party/libevent-1.4.14.fb-changes.diff

## Installation

You may need to point CMake to the location of your custom libcurl and libevent, or to any other libraries which needed to be installed. The *CMAKE_PREFIX_PATH* variable is used to hint to the location.

    export CMAKE_PREFIX_PATH=/home/user

To build HipHop, use the following:

Linux:

    cd /home/user/dev
    git clone git://github.com/facebook/hiphop-php.git
    cd hiphop-php
    git submodule init
    git submodule update
    export HPHP_HOME=`pwd`
    export HPHP_LIB=`pwd`/bin
    cmake .

If you are using FreeBSD instead use export - setenv

Once this is done you can generate the build file. This will return you to the shell. Finally, to build, run `make`. If any errors occur, it may be required to remove the CMakeCache.txt directory in the checkout.

    make

## Contributing to HipHop
HipHop is licensed under the PHP and Zend licenses except as otherwise noted.

Before changes can be accepted a [Contributors Licensing Agreement](http://developers.facebook.com/opensource/cla) must be signed and returned.

## Running HipHop

Please see [the wiki page](http://wiki.github.com/facebook/hiphop-php/running-hiphop)


## Debian Build 
sudo apt-get install debootstrap

mkdir ~/debian_squeeze/

sudo debootstrap squeeze squeeze

sudo mount --bind /proc/ proc/

sudo mount --bind /dev dev

sudo mount --bind /dev/pts dev/pts

sudo chroot .

------ inside of debian


apt-get update

apt-get install gcc

apt-get install build-essential git


git clone git://github.com/h4ck3rm1k3/hiphop-php.git

mv hiphop-php hiphop-php-0.1

apt-get install debhelper cmake libtbb-dev libmcrypt-dev re2c binutils-dev libonig-dev libmysqlclient15-dev libgd2-xpm-dev libmemcached-dev libboost-all-dev libpcre3-dev libevent-dev libboost-all-dev libxml2-dev libbz2-dev libncurses-dev libreadline-dev libc-client2007e-dev libcap-dev  autoconf automake autotools-dev bison curl flex libapache2-mod-php5 libapr1 libaprutil1  libaprutil1-dbd-sqlite3 libaprutil1-ldap libcurl4-openssl-dev libltdl-dev libltdl7 libmhash2 libqdbm14 libssh2-1-dev libtool m4 mcrypt php5-cli php5-common   php5-dev php5-suhosin shtool ssl-cert emacs23-nox

#build the deb
dpkg-buildpackage -uc -us -nc


debian/changelog
hhvm (3.6.5+dfsg1-1+wm8) trusty-wikimedia; urgency=medium

  * Fix broken support for float timeout in socket streaming

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Fri, 29 Jan 2016 11:22:59 +0100

hhvm (3.6.5+dfsg1-1+wm7) trusty-wikimedia; urgency=medium

  * Fix Regex in sql statistics code

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Thu, 01 Oct 2015 12:49:26 +0200

hhvm (3.6.5+dfsg1-1+wm7) trusty-wikimedia; urgency=medium

  * Fix Regex in sql statistics code

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Thu, 01 Oct 2015 12:49:26 +0200

hhvm (3.6.5+dfsg1-1+wm6) trusty-wikimedia; urgency=medium

  * Fix ReflectionClass::getMethods filter
  * filter_var_array: do not fall back to FILTER_DEFAULT
  * Avoid including webscalesqlclient completely
  * Limit log message length for unserialize failures

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Tue, 15 Sep 2015 12:23:50 +0200

hhvm (3.6.5+dfsg1-1+wm5) trusty-wikimedia; urgency=medium

  * Backport fix from https://github.com/facebook/hhvm/pull/6071 (take 3).

 -- Ori Livneh <ori@wikimedia.org>  Thu, 03 Sep 2015 18:09:11 +0000

hhvm (3.6.5+dfsg1-1+wm4) trusty-wikimedia; urgency=medium

  * Backport fix from https://github.com/facebook/hhvm/pull/6071 (take 2).

 -- Ori Livneh <ori@wikimedia.org>  Fri, 21 Aug 2015 00:29:18 +0000

hhvm (3.6.5+dfsg1-1+wm3) trusty-wikimedia; urgency=medium

  * Backport fix from https://github.com/facebook/hhvm/pull/6071.

 -- Ori Livneh <ori@wikimedia.org>  Wed, 19 Aug 2015 22:34:36 +0000

hhvm (3.6.5+dfsg1-1+wm2) trusty-wikimedia; urgency=medium

  * Backport of D40473: Port strtr from zend 5.6.10 (ebernhardson)

 -- Ori Livneh <ori@wikimedia.org>  Fri, 14 Aug 2015 17:28:50 +0000

hhvm (3.6.5+dfsg1-1+wm1) trusty-wikimedia; urgency=medium

  * Upgrade to 3.6.5
  * Session corruption handling patch

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Wed, 22 Jul 2015 14:57:02 +0200

hhvm (3.6.1+dfsg1-1+wm4) trusty-wikimedia; urgency=medium

  * Add a fix for multiple Closure's generated with the same internal name

 -- Erik Bernhardson <ebernhardson@wikimedia.org>  Wed, 17 Jun 2015 22:29:27 -0700

hhvm (3.6.1+dfsg1-1+wm3) trusty-wikimedia; urgency=high

  * Patch for CVE-2015-4024

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Thu, 04 Jun 2015 06:54:47 +0000

hhvm (3.6.1+dfsg1-1+wm2) trusty-wikimedia; urgency=medium

  * Patch for CVE-2015-3413

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Thu, 04 Jun 2015 08:35:46 +0200

hhvm (3.6.1+dfsg1-1+wm1) trusty-wikimedia; urgency=medium

  [ Giuseppe Lavagetto ]
  * New upstream release (3.6.1)
  * Added WMF patch to support streaming output in FastCGI
  * Added patches to build correctly

  [ Alexandros Kosiaris ]
  * Rebuild with hhvm-dev Depends field populated better

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Thu, 16 Apr 2015 12:09:55 +0200

hhvm (3.3.1+dfsg1-1+wm3-1) trusty-wikimedia; urgency=medium

  * debian/patches:
    - Added a patch to allow a new version of the LRU PCRE cache
    - Added a patch to fix a memory leak for bitwise string operations
  * Fixed the upstart/init scripts to use a single ini file

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Thu, 15 Jan 2015 16:04:39 +0100

hhvm (3.3.1+dfsg1-1+wm1) trusty-wikimedia; urgency=medium

  * Upgrade to 3.3.1
  * Add a fix for a libbost memleak on trusty
  * Add a fix for wddx serialization
  * Make HHVM use the system tzinfo
  * remove build-dep alternative for libjpeg62-dev which was confusing trusty

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Wed, 17 Dec 2014 10:46:22 +0100

hhvm (3.3.0+dfsg1-1+wm5) trusty-wikimedia; urgency=medium

  * Add EZC fixes from https://github.com/facebook/hhvm/pull/4330

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Tue, 02 Dec 2014 13:26:32 +0100

hhvm (3.3.0+dfsg1-1+wm4) trusty-wikimedia; urgency=medium

  * Add the /jemalloc-status endpoint to the admin server

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Mon, 10 Nov 2014 13:25:11 +0100

hhvm (3.3.0+dfsg1-1+wm3.1) trusty-wikimedia; urgency=medium

  * Rebuild with RelWithDebInfo

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Fri, 07 Nov 2014 06:28:27 +0100

hhvm (3.3.0+dfsg1-1+wm3) trusty-wikimedia; urgency=medium
  [Wikimedia specific]
  * debian/patches:
    - Added a patch to fix the memleaks and the segfaults with domDocument
    - Added patch to fix the restart behaviour of HHVM

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Thu, 06 Nov 2014 09:57:59 +0100

hhvm (3.3.0+dfsg1-1+wm1) trusty-wikimedia; urgency=medium
  [ Giuseppe Lavagetto ]
  * Updated to version 3.3.0
  * debian/patches:
    - Removed all patches obsoleted by 3.3.0
    - Updated remaining patches so that they apply cleanly
    - Use system-provided pcre library
    - Fixed the config file path that is broken in 3.3.0
    - Backported some patches from upstream for stability/functionality

  [ Wikimedia specific ]
  * Rebased the source on upstream
  * Added WMF-specific patches for PCRE and sweepers init
  * Added the upstart script and updated the init script

 -- Giuseppe Lavagetto <glavagetto@wikimedia.org>  Tue, 21 Oct 2014 12:00:22 +0200

hhvm (3.2.0+dfsg1-2) unstable; urgency=medium

   [ Faidon Liambotis ]
   * Fix the build system to be able to build a release build but with
     debugging symbols (which we subsequently strip into hhvm-dbg), and pass
     -DCMAKE_BUILD_TYPE=RelWithDebInfo to configure.

  [ David Martínez Moreno ]
  * Remove the chmod 750 on /var/log/hhvm as it's really an error on the HHVM
    packaging.
  * debian/patches:
    - disable_quicklz_code: Disable the qlz* primitives, as they are
      GPL-licensed code linked to PHP-licensed one.
    - static_linking_against_libbfd: Static linking against libbfd per
      binutils-dev, backported from HEAD.
    - add_additional_includes_imagemagick: New ImageMagick broke the build,
      so add the arch includes to the build.
  * debian/copyright: Fixed some mistakes discovered with latest lintian.

src/util/hash.h
/*
   +----------------------------------------------------------------------+
   | HipHop for PHP                                                       |
   +----------------------------------------------------------------------+
   | Copyright (c) 2010 Facebook, Inc. (http://www.facebook.com)          |
   +----------------------------------------------------------------------+
   | This source file is subject to version 3.01 of the PHP license,      |
   | that is bundled with this package in the file LICENSE, and is        |
   | available through the world-wide-web at the following url:           |
   | http://www.php.net/license/3_01.txt                                  |
   | If you did not receive a copy of the PHP license and are unable to   |
   | obtain it through the world-wide-web, please send a note to          |
   | license@php.net so we can mail you a copy immediately.               |
   +----------------------------------------------------------------------+
*/

#ifndef __HPHP_HASH_H__
#define __HPHP_HASH_H__

namespace HPHP {
///////////////////////////////////////////////////////////////////////////////

inline long long hash_int64(long long key) {
  key = (~key) + (key << 21); // key = (key << 21) - key - 1;
  key = key ^ ((unsigned long long)key >> 24);
  key = (key + (key << 3)) + (key << 8); // key * 265
  key = key ^ ((unsigned long long)key >> 14);
  key = (key + (key << 2)) + (key << 4); // key * 21
  key = key ^ ((unsigned long long)key >> 28);
  key = key + (key << 31);
  return key < 0 ? -key : key;
}

/*
 * How to toggle hash functions: comment/uncomment the following macro
 * definition, build hphp, make -C lib/system, and then rebuild hphp.
 */
#define USE_MURMUR 1

#ifdef USE_MURMUR

/*
 * http://murmurhash.googlepages.com/ (64bit, seed = 0)
 *
 * The case-insensitive version converts 8 bytes of lowercased characters
 * uppercase at once. This should work as identifiers usually only contain
 * alphanumeric characters and the underscore. Although PHP allows higher
 * ASCII characters (> 127) in an identifier, they should be very rare, and
 * do not change the correctness.
 */

inline long long hash_string(const char *arKey, int nKeyLength) {
  const long long m = 0xc6a4a7935bd1e995ll;
  const int r = 47;

  register unsigned long long h = 0;

  const unsigned long long * data = (const unsigned long long *)arKey;
  const unsigned long long * end = data + (nKeyLength / 8);

  while (data != end) {
    unsigned long long k = *data++;

    k *= m;
    k ^= k >> r;
    k *= m;

    h ^= k;
    h *= m;
  }

  const unsigned char * data2 = (const unsigned char*)data;

  switch (nKeyLength & 7) {
  case 7: h ^= (unsigned long long)(data2[6]) << 48;
  case 6: h ^= (unsigned long long)(data2[5]) << 40;
  case 5: h ^= (unsigned long long)(data2[4]) << 32;
  case 4: h ^= (unsigned long long)(data2[3]) << 24;
  case 3: h ^= (unsigned long long)(data2[2]) << 16;
  case 2: h ^= (unsigned long long)(data2[1]) << 8;
  case 1: h ^= (unsigned long long)(data2[0]);
          h *= m;
  };

  h ^= h >> r;
  h *= m;
  h ^= h >> r;

  return h & 0x7fffffffffffffffll;
}

inline long long hash_string_i(const char *arKey, int nKeyLength) {
  const unsigned long long m = 0xc6a4a7935bd1e995ll;
  const int r = 47;

  register unsigned long long h = 0;

  const unsigned long long * data = (const unsigned long long *)arKey;
  const unsigned long long * end = data + (nKeyLength / 8);

  while (data != end) {
    unsigned long long k = *data++;
    k &= 0xdfdfdfdfdfdfdfdfll; // a-z => A-Z

    k *= m;
    k ^= k >> r;
    k *= m;

    h ^= k;
    h *= m;
  }

  const unsigned char * data2 = (const unsigned char*)data;

  switch (nKeyLength & 7) {
  case 7: h ^= (unsigned long long)(data2[6] & 0xdf) << 48;
  case 6: h ^= (unsigned long long)(data2[5] & 0xdf) << 40;
  case 5: h ^= (unsigned long long)(data2[4] & 0xdf) << 32;
  case 4: h ^= (unsigned long long)(data2[3] & 0xdf) << 24;
  case 3: h ^= (unsigned long long)(data2[2] & 0xdf) << 16;
  case 2: h ^= (unsigned long long)(data2[1] & 0xdf) << 8;
  case 1: h ^= (unsigned long long)(data2[0] & 0xdf);
          h *= m;
  };

  h ^= h >> r;
  h *= m;
  h ^= h >> r;

  return h & 0x7fffffffffffffffll;
}

#else /* not using murmur hashing */

/*
 * DJBX33A (Daniel J. Bernstein, Times 33 with Addition)
 *
 * This is Daniel J. Bernstein's popular `times 33' hash function as
 * posted by him years ago on comp.lang.c. It basically uses a function
 * like ``hash(i) = hash(i-1) * 33 + str[i]''. This is one of the best
 * known hash functions for strings. Because it is both computed very
 * fast and distributes very well.
 *
 * The magic of number 33, i.e. why it works better than many other
 * constants, prime or not, has never been adequately explained by
 * anyone. So I try an explanation: if one experimentally tests all
 * multipliers between 1 and 256 (as RSE did now) one detects that even
 * numbers are not useable at all. The remaining 128 odd numbers
 * (except for the number 1) work more or less all equally well. They
 * all distribute in an acceptable way and this way fill a hash table
 * with an average percent of approx. 86%.
 *
 * If one compares the Chi^2 values of the variants, the number 33 not
 * even has the best value. But the number 33 and a few other equally
 * good numbers like 17, 31, 63, 127 and 129 have nevertheless a great
 * advantage to the remaining numbers in the large set of possible
 * multipliers: their multiply operation can be replaced by a faster
 * operation based on just one shift plus either a single addition
 * or subtraction operation. And because a hash function has to both
 * distribute good _and_ has to be very fast to compute, those few
 * numbers should be preferred and seems to be the reason why Daniel J.
 * Bernstein also preferred it.
 *
 *
 *                  -- Ralf S. Engelschall <rse@engelschall.com>
 */

inline long long hash_string(const char *arKey, int nKeyLength) {
  register unsigned long long hash = 5381;

  /* variant with the hash unrolled eight times */
  for (; nKeyLength >= 8; nKeyLength -= 8) {
    hash = ((hash << 5) + hash) + *arKey++;
    hash = ((hash << 5) + hash) + *arKey++;
    hash = ((hash << 5) + hash) + *arKey++;
    hash = ((hash << 5) + hash) + *arKey++;
    hash = ((hash << 5) + hash) + *arKey++;
    hash = ((hash << 5) + hash) + *arKey++;
    hash = ((hash << 5) + hash) + *arKey++;
    hash = ((hash << 5) + hash) + *arKey++;
  }
  switch (nKeyLength) {
  case 7: hash = ((hash << 5) + hash) + *arKey++; /* fallthrough... */
  case 6: hash = ((hash << 5) + hash) + *arKey++; /* fallthrough... */
  case 5: hash = ((hash << 5) + hash) + *arKey++; /* fallthrough... */
  case 4: hash = ((hash << 5) + hash) + *arKey++; /* fallthrough... */
  case 3: hash = ((hash << 5) + hash) + *arKey++; /* fallthrough... */
  case 2: hash = ((hash << 5) + hash) + *arKey++; /* fallthrough... */
  case 1: hash = ((hash << 5) + hash) + *arKey++; break;
  case 0: break;
  default:
    ASSERT(false);
    break;
  }
  long long ret = hash;
  return ret < 0 ? -ret : ret;
}

// Case insensitive version. Hash will be equivalent to hash_string of lower
inline long long hash_string_i(const char *arKey, int nKeyLength) {
  register unsigned long long hash = 5381;
  char c;
  char diff = 'a' - 'A';
  /* variant with the hash unrolled eight times */
  for (; nKeyLength >= 8; nKeyLength -= 8) {
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
  }
  switch (nKeyLength) {
  case 7:
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
  case 6:
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
  case 5:
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
  case 4:
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
  case 3:
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
  case 2:
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
  case 1:
    c = *arKey++;
    if (c >= 'A' && c <= 'Z') c += diff;
    hash = ((hash << 5) + hash) + c;
  case 0: break;
  default:
    ASSERT(false);
    break;
  }
  long long ret = hash;
  return ret < 0 ? -ret : ret;
}

#endif /* USE_MURMUR */

/**
 * We probably should get rid of this, so to detect code generation errors,
 * where a binary string is treated as a NULL-terminated literal. Do we ever
 * allow binary strings as array keys or symbol names?
 */
inline long long hash_string(const char *arKey) {
  return hash_string(arKey, strlen(arKey));
}
inline long long hash_string_i(const char *arKey) {
  return hash_string_i(arKey, strlen(arKey));
}

inline bool is_strictly_integer(const char* arKey, size_t nKeyLength,
                                int64& res) {
  if (nKeyLength > 0 &&
      (nKeyLength <= 19 ||
       (arKey[0] == '-' && nKeyLength == 20))) {
    uint64 num = 0;
    bool neg = false;
    uint32 i = 0;
    if (arKey[0] == '-') {
      neg = true;
      i = 1;
      if (nKeyLength == 1) return false;
    }
    if (nKeyLength == i + 1 || arKey[i] != '0') {
      bool good = true;
      for (; i < nKeyLength; ++i) {
        if (arKey[i] >= '0' && arKey[i] <= '9') {
          num = 10*num + arKey[i] - '0';
        }
        else {
          good = false;
          break;
        }
      }
      if (good && num <= 0x7FFFFFFFFFFFFFFFll) {
        res = neg ? 0 - num : (int64)num;
        return true;
      }
    }
  }
  return false;
}

///////////////////////////////////////////////////////////////////////////////
}

#endif // __HPHP_HASH_H__
